lib/puppet/type/cert_file.rb

```diff
@@ -0,0 +1,45 @@
+Puppet::Type.newtype(:cert_file) do
+  @doc = 'Manages X.509 certificate files downloaded from a source location, saved in the specified format.'
+
+  ensurable
+
+  newparam(:path, namevar: true) do
+    desc 'Path to the file to manage'
+    validate do |value|
+      unless Puppet::Util.absolute_path?(value)
+        raise Puppet::Error, _("File paths must be fully qualified, not '%{path}'") % { path: value }
+      end
+    end
+
+    munge do |value|
+      if value.start_with?('//') && ::File.basename(value) == '/'
+        # This is a UNC path pointing to a share, so don't add a trailing slash
+        File.expand_path(value)
+      else
+        File.join(::File.split(::File.expand_path(value)))
+      end
+    end
+  end # newparam(:path)
+
+  newparam(:source) do
+    validate do |source|
+      begin
+        uri = URI.parse(Puppet::Util.uri_encode(source))
+      rescue => detail
+        raise Puppet::Error, "Could not understand source #{source}: #{detail}", detail
```

Comment:
Why is there a third parameter?

Reply:
To throw the original exception for backtrace.